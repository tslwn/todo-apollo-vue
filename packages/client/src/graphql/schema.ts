export interface Todo {
  __typename: 'Todo';
  id: number;
  text: string;
  isComplete: boolean;
  isArchived: boolean;
  createdAt?: string;
  updatedAt?: string;
}

export enum Sort {
  Asc = 'ASC',
  Desc = 'DESC',
}

export type TodosFilterInput = Partial<Pick<Todo, 'isComplete' | 'isArchived'>>;

export type TodosOrderByInputKey = keyof Pick<Todo, 'createdAt'>;

export type TodosOrderByInput = {
  [K in TodosOrderByInputKey]?: Sort;
};

export interface TodosInput {
  filter?: TodosFilterInput;
  orderBy?: TodosOrderByInput;
}

export type TodoInput = Pick<Todo, 'id'>;

export interface Query {
  __typename: 'Query';
}

export interface TodosResponse extends Query {
  todos: Todo[];
}

export type AddTodoInput = Pick<Todo, 'text'>;

export type ChangeTodoTextInput = Pick<Todo, 'id' | 'text'>;

export type ChangeTodoIsCompleteInput = Pick<Todo, 'id' | 'isComplete'>;

export type ChangeTodoIsArchivedInput = Pick<Todo, 'id' | 'isArchived'>;

interface TodoUpdateResponseGeneric {
  __typename: 'TodoUpdateResponse';
  success: boolean;
  message?: string;
}

interface TodoUpdateResponseSuccess extends TodoUpdateResponseGeneric {
  success: true;
  todo: Pick<Todo, '__typename' | 'id' | 'text' | 'isComplete' | 'isArchived'>;
}

interface TodoUpdateResponseFailure extends TodoUpdateResponseGeneric {
  success: false;
}

export type TodoUpdateResponse =
  | TodoUpdateResponseSuccess
  | TodoUpdateResponseFailure;

export interface Mutation {
  __typename: 'Mutation';
}

export interface AddTodoResponse extends Mutation {
  addTodo: TodoUpdateResponse;
}

export interface ChangeTodoTextResponse extends Mutation {
  changeTodoText: TodoUpdateResponse;
}

export interface ChangeTodoIsCompleteResponse extends Mutation {
  changeTodoIsComplete: TodoUpdateResponse;
}

export interface ChangeTodoIsArchivedResponse extends Mutation {
  changeTodoIsArchived: TodoUpdateResponse;
}
