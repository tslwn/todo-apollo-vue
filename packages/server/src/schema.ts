import { gql } from 'apollo-server';
import { Todo, TodoAttributes } from 'types/todo.types';

const schema = gql`
  type Todo {
    id: ID!
    text: String!
    isComplete: Boolean!
    isArchived: Boolean!
    createdAt: String!
    updatedAt: String!
  }

  enum Sort {
    ASC
    DESC
  }

  input TodoFilterInput {
    isComplete: Boolean
    isArchived: Boolean
  }

  input TodoOrderByInput {
    createdAt: Sort
  }

  type Query {
    todos(filter: TodoFilterInput, orderBy: TodoOrderByInput): [Todo]!
    todo(id: ID!): Todo
  }

  type TodoUpdateResponse {
    success: Boolean!
    message: String
    todo: Todo
  }

  type Mutation {
    addTodo(text: String!): TodoUpdateResponse!
    changeTodoText(id: ID!, text: String!): TodoUpdateResponse!
    changeTodoIsComplete(id: ID!, isComplete: Boolean!): TodoUpdateResponse!
    changeTodoIsArchived(id: ID!, isArchived: Boolean!): TodoUpdateResponse!
  }
`;

export enum Sort {
  ASC,
  DESC
}

export type TodosFilterInput = Partial<Pick<Todo, 'isComplete' | 'isArchived'>>;

type TodosOrderByInputKey = keyof Pick<Todo, 'createdAt'>;

export type TodosOrderByInput = {
  [index: string]: { [K in TodosOrderByInputKey]?: Sort };
};

export interface TodosInput {
  filter?: TodosFilterInput;
  orderBy?: TodosOrderByInput;
}

export type TodoInput = Pick<Todo, 'id'>;

export type AddTodoInput = Pick<Todo, 'text'>;

export type ChangeTodoTextInput = Pick<Todo, 'id' | 'text'>;

export type ChangeTodoIsCompleteInput = Pick<Todo, 'id' | 'isComplete'>;

export type ChangeTodoIsArchivedInput = Pick<Todo, 'id' | 'isArchived'>;

export interface TodoUpdateResponse {
  success: boolean;
  message?: string;
  todo?: Todo;
}

export default schema;
