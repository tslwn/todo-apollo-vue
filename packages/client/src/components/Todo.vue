<template>
  <v-list-item>
    <v-list-item-action>
      <v-checkbox v-model="todo.isComplete" @change="onIsCompleteChange" />
    </v-list-item-action>
    <v-list-item-content>
      <v-list-item-title v-text="todo.text"></v-list-item-title>
    </v-list-item-content>
    <v-list-item-icon>
      <v-btn icon @click="onArchiveClick">
        <v-icon>{{ todo.isArchived ? 'mdi-delete-off' : 'mdi-delete' }}</v-icon>
      </v-btn>
    </v-list-item-icon>
  </v-list-item>
</template>
<script lang="ts">
import Vue, { PropType } from 'vue';
import {
  changeTodoIsArchivedMutation,
  changeTodoIsCompleteMutation,
  todosQuery,
} from '../graphql';
import { TODOS_VARIABLES } from './TodoList.vue';

// export from somewhere?
interface Todo {
  id: number;
  text: string;
  isComplete: boolean;
  isArchived: boolean;
}

export default Vue.extend({
  name: 'Todo',
  props: {
    todo: {
      type: Object as PropType<Todo>,
      required: true,
    },
  },
  methods: {
    onIsCompleteChange(isComplete: boolean) {
      const { id } = this.todo;

      this.$apollo
        .mutate({
          mutation: changeTodoIsCompleteMutation,
          variables: {
            id,
            isComplete,
          },
        })
        .catch(() => {
          this.todo.isComplete = !this.todo.isComplete;
        });
    },
    onArchiveClick() {
      const { id, text, isComplete, isArchived } = this.todo;

      this.$apollo
        .mutate({
          mutation: changeTodoIsArchivedMutation,
          variables: {
            id,
            isArchived: !isArchived,
          },
          update: (store, { data: { changeTodoIsArchived } }) => {
            const { message, success, todo } = changeTodoIsArchived;

            if (!success) {
              throw new Error(message);
            }

            const data = store.readQuery({
              query: todosQuery,
              variables: TODOS_VARIABLES,
            });

            // remove todo from query cache if not included by filter
            const todoInFilter =
              TODOS_VARIABLES.filter.isArchived === undefined ||
              TODOS_VARIABLES.filter.isArchived === todo.isArchived;

            if (!todoInFilter) {
              store.writeQuery({
                data: {
                  todos: data.todos.filter((item: Todo) => item.id !== todo.id),
                },
                query: todosQuery,
                variables: TODOS_VARIABLES,
              });
            }
          },
          optimisticResponse: {
            __typename: 'Mutation',
            changeTodoIsArchived: {
              message: 'Todo isArchived changed successfully',
              success: true,
              todo: {
                __typename: 'Todo',
                id,
                text,
                isComplete,
                isArchived: !isArchived,
              },
              __typename: 'TodoUpdateResponse',
            },
          },
        })
        .catch(() => {
          // restore value
          this.todo.isArchived = isArchived;
        });
    },
  },
});
</script>
<style scoped></style>
