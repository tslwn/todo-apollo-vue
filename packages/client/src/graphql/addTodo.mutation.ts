import gql from 'graphql-tag';
import { MutationUpdaterFn } from 'apollo-boost';
import { AddTodoResponse, TodosResponse } from './schema';
import { todosQuery, TODOS_VARIABLES } from './todos.query';

export const addTodoMutation = gql`
  mutation($text: String!) {
    addTodo(text: $text) {
      success
      message
      todo {
        id
        text
        isComplete
        isArchived
      }
    }
  }
`;

export const addTodoUpdate: MutationUpdaterFn<AddTodoResponse> = (
  store,
  mutationResult,
) => {
  const { data } = mutationResult;
  if (!data) {
    throw new Error('No mutation result');
  }

  const { message, success, todo } = data.addTodo;
  if (!success) {
    throw new Error(message);
  }

  const queryResponse = store.readQuery<TodosResponse>({
    query: todosQuery,
    variables: TODOS_VARIABLES,
  });
  if (queryResponse === null) {
    throw new Error('No query result');
  }

  // add todo to cache
  const todos =
    TODOS_VARIABLES?.orderBy?.createdAt === 'DESC'
      ? [todo, ...queryResponse.todos]
      : [...queryResponse.todos, todo];

  store.writeQuery({
    data: {
      todos,
    },
    query: todosQuery,
    variables: TODOS_VARIABLES,
  });
};
