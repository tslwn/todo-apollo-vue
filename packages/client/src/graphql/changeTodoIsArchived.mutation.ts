import gql from 'graphql-tag';
import { MutationUpdaterFn } from 'apollo-boost';
import { Todo, ChangeTodoIsArchivedResponse, TodosResponse } from './schema';
import { todosQuery, TODOS_VARIABLES } from './todos.query';

export const changeTodoIsArchivedMutation = gql`
  mutation($id: ID!, $isArchived: Boolean!) {
    changeTodoIsArchived(id: $id, isArchived: $isArchived) {
      success
      message
      todo {
        id
        text
        isComplete
        isArchived
      }
    }
  }
`;

export const changeTodoIsArchivedUpdate: MutationUpdaterFn<ChangeTodoIsArchivedResponse> = (
  store,
  mutationResult,
) => {
  const { data } = mutationResult;
  if (data == null) {
    throw new Error('No mutation result');
  } else if (data.changeTodoIsArchived.success === false) {
    throw new Error(data.changeTodoIsArchived.message);
  } else {
    const todo = data.changeTodoIsArchived.todo;

    // get todos before update
    const queryResponse = store.readQuery<TodosResponse>({
      query: todosQuery,
      variables: TODOS_VARIABLES,
    });
    if (queryResponse === null) {
      throw new Error('No query result');
    }

    // remove todo from query cache if not included by filter
    const todoInFilter =
      TODOS_VARIABLES?.filter?.isArchived === undefined ||
      TODOS_VARIABLES.filter.isArchived === todo.isArchived;

    if (!todoInFilter) {
      store.writeQuery({
        data: {
          todos: queryResponse.todos.filter(
            (item: Todo) => item.id !== todo?.id,
          ),
        },
        query: todosQuery,
        variables: TODOS_VARIABLES,
      });
    }
  }
};

/**
 * TODO: distinguish between 'attributes' and 'creation attributes' following
 * server package `src/types/todo.types`
 */
export const changeTodoIsArchivedOptimisticResponse = ({
  id,
  text,
  isComplete,
  isArchived,
}: Pick<
  Todo,
  'id' | 'text' | 'isComplete' | 'isArchived'
>): ChangeTodoIsArchivedResponse => ({
  __typename: 'Mutation',
  changeTodoIsArchived: {
    message: 'Todo isArchived changed successfully',
    success: true,
    todo: {
      __typename: 'Todo',
      id,
      text,
      isComplete,
      isArchived: !isArchived,
    },
    __typename: 'TodoUpdateResponse',
  },
});
