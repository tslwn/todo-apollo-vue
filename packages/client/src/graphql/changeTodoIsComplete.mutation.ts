import gql from 'graphql-tag';

export default gql`
  mutation($id: ID!, $isComplete: Boolean!) {
    changeTodoIsComplete(id: $id, isComplete: $isComplete) {
      success
      message
      todo {
        id
        text
        isComplete
        isArchived
      }
    }
  }
`;
