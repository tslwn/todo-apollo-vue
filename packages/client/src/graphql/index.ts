import addTodoMutation from './addTodo.mutation';
import changeTodoIsArchivedMutation from './changeTodoIsArchived.mutation';
import changeTodoIsCompleteMutation from './changeTodoIsComplete.mutation';
import todosQuery from './todos.query';

export {
  addTodoMutation,
  changeTodoIsArchivedMutation,
  changeTodoIsCompleteMutation,
  todosQuery,
};
