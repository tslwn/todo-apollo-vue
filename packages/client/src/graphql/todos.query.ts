import gql from 'graphql-tag';

export default gql`
  query($filter: TodoFilterInput, $orderBy: TodoOrderByInput) {
    todos(filter: $filter, orderBy: $orderBy) {
      id
      text
      isComplete
      isArchived
    }
  }
`;
