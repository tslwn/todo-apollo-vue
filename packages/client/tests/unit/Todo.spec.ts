import { graphql } from 'graphql';
import { addMocksToSchema } from '@graphql-tools/mock';
import Vue from 'vue';
import Vuetify from 'vuetify';
import { mount, createLocalVue, ThisTypedMountOptions } from '@vue/test-utils';
import {
  changeTodoIsArchivedMutation,
  changeTodoIsArchivedUpdate,
  changeTodoIsArchivedOptimisticResponse,
} from '../../src/graphql/changeTodoIsArchived.mutation';
import changeTodoIsCompleteMutation from '../../src/graphql/changeTodoIsComplete.mutation';
import Todo from '../../src/components/Todo.vue';
import schema from '../mockSchema';

Vue.use(Vuetify);

const localVue = createLocalVue();

describe('Todo.vue', () => {
  let vuetify: typeof Vuetify;

  const createComponent = (options?: ThisTypedMountOptions<Vue>) => {
    return mount(Todo, {
      localVue,
      vuetify,
      ...options,
    });
  };

  beforeEach(() => {
    vuetify = new Vuetify();
    addMocksToSchema({ schema, preserveResolvers: true });
  });

  /**
   * Snapshot tests. Note that because isComplete and isArchived are independent
   * (the checkbox and icon button could be moved to child components) there is
   * no need to test the two remaining permutations.
   */

  const todo = {
    __typename: 'Todo' as const,
    id: -1,
    text: 'Add unit tests',
    isComplete: false,
    isArchived: false,
  };

  it('renders correctly with isComplete false, isArchived false', () => {
    const wrapper = createComponent({
      propsData: { todo },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with isComplete true, isArchived true', () => {
    const wrapper = createComponent({
      propsData: { todo: { ...todo, isArchived: true, isComplete: true } },
    });
    expect(wrapper).toMatchSnapshot();
  });

  /**
   * User interaction tests. TODO: trigger events instead of calling methods
   * directly. I'm not sure whether it would be best to combine the two steps.
   */

  it('onIsCompleteChange method calls Apollo mutation', () => {
    const mutate = jest.fn().mockImplementation(() => ({ catch: jest.fn() }));
    const wrapper = createComponent({
      mocks: {
        $apollo: {
          mutate,
        },
      },
      propsData: {
        todo,
      },
    });
    const isComplete = !todo.isComplete;
    (wrapper.vm as any).onIsCompleteChange(isComplete);
    expect(mutate).toHaveBeenCalledWith({
      mutation: changeTodoIsCompleteMutation,
      variables: {
        id: todo.id,
        isComplete,
      },
    });
  });

  it('onArchiveClick method calls Apollo mutation', () => {
    const mutate = jest.fn().mockImplementation(() => ({ catch: jest.fn() }));
    const wrapper = createComponent({
      mocks: {
        $apollo: {
          mutate,
        },
      },
      propsData: {
        todo,
      },
    });
    (wrapper.vm as any).onArchiveClick();
    expect(mutate).toHaveBeenCalledWith({
      mutation: changeTodoIsArchivedMutation,
      variables: {
        id: todo.id,
        isArchived: !todo.isArchived,
      },
      update: changeTodoIsArchivedUpdate,
      optimisticResponse: changeTodoIsArchivedOptimisticResponse(todo),
    });
    /**
     * TODO: verify cache after `update` called. Could perhaps also check it's
     * called twice, first with the optimistic response and then the presumed
     * actual response.
     */
  });

  /**
   * Mock schema tests.
   */

  it('called changeTodoIsComplete mutation with mock schema', async () => {
    const response = {
      success: true,
      message: 'Todo isComplete changed successfully',
      todo: {
        id: 0,
        text: 'Add unit tests',
        isComplete: true,
        isArchived: false,
      },
    };
    const mutation = `
      mutation($id: ID!, $isComplete: Boolean!) {
        changeTodoIsComplete(id: $id, isComplete: $isComplete) {
          success
          message
          todo {
            id
            text
            isComplete
            isArchived
          }
        }
      }
    `;
    const result = await graphql({
      schema,
      source: mutation,
      variableValues: {
        id: response.todo.id,
        isComplete: response.todo.isComplete,
      },
    });
    if (result.data == null) {
      throw new Error('No mutation result');
    }
    expect(result.data.changeTodoIsComplete).toMatchObject(response);
  });

  it('called changeTodoIsArchived mutation with mock schema', async () => {
    const response = {
      success: true,
      message: 'Todo isArchived changed successfully',
      todo: {
        id: 0,
        text: 'Add unit tests',
        isComplete: false,
        isArchived: true,
      },
    };
    const mutation = `
      mutation($id: ID!, $isArchived: Boolean!) {
        changeTodoIsArchived(id: $id, isArchived: $isArchived) {
          success
          message
          todo {
            id
            text
            isComplete
            isArchived
          }
        }
      }
    `;
    const result = await graphql({
      schema,
      source: mutation,
      variableValues: {
        id: response.todo.id,
        isArchived: response.todo.isArchived,
      },
    });
    if (result.data == null) {
      throw new Error('No mutation result');
    }
    expect(result.data.changeTodoIsArchived).toMatchObject(response);
  });
});
