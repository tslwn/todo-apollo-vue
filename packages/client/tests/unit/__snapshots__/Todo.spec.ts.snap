// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo.vue renders correctly with isComplete false, isArchived false 1`] = `
<div tabindex="-1" class="v-list-item theme--light">
  <div class="v-list-item__action">
    <div class="v-input theme--light v-input--selection-controls v-input--checkbox">
      <div class="v-input__control">
        <div class="v-input__slot">
          <div class="v-input--selection-controls__input"><i aria-hidden="true" class="v-icon notranslate mdi mdi-checkbox-blank-outline theme--light"></i><input aria-checked="false" id="input-3" role="checkbox" type="checkbox" value="">
            <div class="v-input--selection-controls__ripple"></div>
          </div>
        </div>
        <div class="v-messages theme--light">
          <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
        </div>
      </div>
    </div>
  </div>
  <div class="v-list-item__content">
    <div class="v-list-item__title">Add unit tests</div>
  </div>
  <div class="v-list-item__icon"><button type="button" class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-delete theme--light"></i></span></button></div>
</div>
`;

exports[`Todo.vue renders correctly with isComplete true, isArchived true 1`] = `
<div tabindex="-1" class="v-list-item theme--light">
  <div class="v-list-item__action">
    <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-input--selection-controls v-input--checkbox primary--text">
      <div class="v-input__control">
        <div class="v-input__slot">
          <div class="v-input--selection-controls__input"><i aria-hidden="true" class="v-icon notranslate mdi mdi-checkbox-marked theme--light primary--text"></i><input aria-checked="true" id="input-12" role="checkbox" type="checkbox" value="">
            <div class="v-input--selection-controls__ripple primary--text"></div>
          </div>
        </div>
        <div class="v-messages theme--light primary--text">
          <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
        </div>
      </div>
    </div>
  </div>
  <div class="v-list-item__content">
    <div class="v-list-item__title">Add unit tests</div>
  </div>
  <div class="v-list-item__icon"><button type="button" class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-delete-off theme--light"></i></span></button></div>
</div>
`;
